From 836f0ab23bf50a7f17f14625e183f6b1bf6f49de Mon Sep 17 00:00:00 2001
From: Xiaoning Ma <mxiaonin@qti.qualcomm.com>
Date: Thu, 26 Mar 2026 06:46:45 +0530
Subject: [PATCH] Bluetooth: qca: add support for QCA automotive BT chips

This patch adds support for QCA Bluetooth chips used on automotive platforms.

Key updates include:

1. BT chip power sequencing is handled by the out-of-tree btpower kernel
   module.
2. IBS and deep sleep features are not supported on automotive platforms.
3. btqca detects QCA chips via soc_id and loads the corresponding firmware
   files.
4. Correct the NVM filename format to use four characters after the prefix 'b'
   instead of two.

Change-Id: Idb24238e36747cbc8ec69f5dc9f6b56163495267
Signed-off-by: Xiaoning Ma <mxiaonin@qti.qualcomm.com>
---
 drivers/bluetooth/btqca.c   | 60 ++++++++++++++++++++++++++++++++++---
 drivers/bluetooth/btqca.h   |  5 ++++
 drivers/bluetooth/hci_qca.c | 29 +++++++++++++++++-
 3 files changed, 89 insertions(+), 5 deletions(-)

diff --git a/drivers/bluetooth/btqca.c b/drivers/bluetooth/btqca.c
index 7e8a05591a85..d7556de1e305 100644
--- a/drivers/bluetooth/btqca.c
+++ b/drivers/bluetooth/btqca.c
@@ -442,7 +442,12 @@ static int qca_tlv_check_data(struct hci_dev *hdev,
 				 * enabling software inband sleep
 				 * onto controller side.
 				 */
-				tlv_nvm->data[0] |= 0x80;
+				if (soc_type == QCA_AUTO_CONVERGED) {
+					BT_ERR("Disable SW IBS");
+					tlv_nvm->data[0] &= ~0x80;
+				} else {
+					tlv_nvm->data[0] |= 0x80;
+				}
 
 				/* UART Baud Rate */
 				if (soc_type >= QCA_WCN3991)
@@ -459,7 +464,13 @@ static int qca_tlv_check_data(struct hci_dev *hdev,
 				/* Sleep enable mask
 				 * enabling deep sleep feature on controller.
 				 */
-				tlv_nvm->data[0] |= 0x01;
+				if (soc_type == QCA_AUTO_CONVERGED) {
+					BT_ERR("Disable deep sleep");
+					tlv_nvm->data[0] = 0;
+					tlv_nvm->data[1] = 0;
+				} else {
+					tlv_nvm->data[0] |= 0x01;
+				}
 
 				break;
 			}
@@ -772,7 +783,7 @@ static void qca_get_nvm_name_by_board(char *fwname, size_t max_size,
 		if (bid == 0x0 || bid == 0xffff)
 			snprintf(fwname, max_size, "qca/%s%02x%s.bin", stem, rom_ver, variant);
 		else
-			snprintf(fwname, max_size, "qca/%s%02x%s.%s%02x", stem, rom_ver,
+			snprintf(fwname, max_size, "qca/%s%02x%s.%s%04x", stem, rom_ver,
 						variant, prefix, bid);
 	} else {
 		if (bid == 0x0 || bid == 0xffff)
@@ -853,6 +864,23 @@ int qca_uart_setup(struct hci_dev *hdev, uint8_t baudrate,
 			snprintf(config.fwname, sizeof(config.fwname),
 				 "qca/hmtbtfw%02x.tlv", rom_ver);
 			break;
+		case QCA_AUTO_CONVERGED:
+			if (ver.soc_id == QCA_QCA6595_SOC_ID)
+				snprintf(config.fwname, sizeof(config.fwname),
+					 "qca/gnbtfw%02x.tlv", rom_ver);
+			else if (ver.soc_id == QCA_QCA6696_SOC_ID)
+				snprintf(config.fwname, sizeof(config.fwname),
+					 "qca/htbtfw%02x.tlv", rom_ver);
+			else if (ver.soc_id == QCA_QCA6698_SOC_ID)
+				snprintf(config.fwname, sizeof(config.fwname),
+					 "qca/hpbtfw%02x.tlv", rom_ver);
+			else if(ver.soc_id == QCA_QCA6797_SOC_ID)
+				snprintf(config.fwname, sizeof(config.fwname),
+					 "qca/hmtbtfw%02x.tlv", rom_ver);
+			else
+				snprintf(config.fwname, sizeof(config.fwname),
+					 "qca/rampatch_%08x.bin", soc_ver);
+			break;
 		default:
 			snprintf(config.fwname, sizeof(config.fwname),
 				 "qca/rampatch_%08x.bin", soc_ver);
@@ -868,7 +896,8 @@ int qca_uart_setup(struct hci_dev *hdev, uint8_t baudrate,
 	/* Give the controller some time to get ready to receive the NVM */
 	msleep(10);
 
-	if (soc_type == QCA_QCA2066 || soc_type == QCA_WCN7850)
+	if (soc_type == QCA_QCA2066 || soc_type == QCA_WCN7850
+	    || soc_type == QCA_AUTO_CONVERGED)
 		qca_read_fw_board_id(hdev, &boardid);
 
 	/* Download NVM configuration */
@@ -919,6 +948,27 @@ int qca_uart_setup(struct hci_dev *hdev, uint8_t baudrate,
 			qca_get_nvm_name_by_board(config.fwname, sizeof(config.fwname),
 				 "hmtnv", soc_type, ver, rom_ver, boardid);
 			break;
+		case QCA_AUTO_CONVERGED:
+			if (ver.soc_id == QCA_QCA6595_SOC_ID)
+				qca_get_nvm_name_by_board(config.fwname,
+					sizeof(config.fwname), "gnnv", soc_type, ver,
+					rom_ver, boardid);
+			else if (ver.soc_id == QCA_QCA6696_SOC_ID)
+				qca_get_nvm_name_by_board(config.fwname,
+					sizeof(config.fwname), "htnv", soc_type, ver,
+					rom_ver, boardid);
+			else if (ver.soc_id == QCA_QCA6698_SOC_ID)
+				qca_get_nvm_name_by_board(config.fwname,
+					sizeof(config.fwname), "hpnv", soc_type, ver,
+					rom_ver, boardid);
+			else if (ver.soc_id == QCA_QCA6797_SOC_ID)
+				qca_get_nvm_name_by_board(config.fwname,
+					sizeof(config.fwname), "hmtnv", soc_type, ver,
+					rom_ver, boardid);
+			else
+				snprintf(config.fwname, sizeof(config.fwname),
+					 "qca/nvm_%08x.bin", soc_ver);
+			break;
 		default:
 			snprintf(config.fwname, sizeof(config.fwname),
 				 "qca/nvm_%08x.bin", soc_ver);
@@ -938,6 +988,7 @@ int qca_uart_setup(struct hci_dev *hdev, uint8_t baudrate,
 	case QCA_WCN6750:
 	case QCA_WCN6855:
 	case QCA_WCN7850:
+	case QCA_AUTO_CONVERGED:
 		err = qca_disable_soc_logging(hdev);
 		if (err < 0)
 			return err;
@@ -973,6 +1024,7 @@ int qca_uart_setup(struct hci_dev *hdev, uint8_t baudrate,
 	case QCA_WCN6750:
 	case QCA_WCN6855:
 	case QCA_WCN7850:
+	case QCA_AUTO_CONVERGED:
 		/* get fw build info */
 		err = qca_read_fw_build_info(hdev);
 		if (err < 0)
diff --git a/drivers/bluetooth/btqca.h b/drivers/bluetooth/btqca.h
index d00b42deccbe..8f522844a32f 100644
--- a/drivers/bluetooth/btqca.h
+++ b/drivers/bluetooth/btqca.h
@@ -40,6 +40,10 @@
 #define QCA_HCI_CC_SUCCESS		0x00
 
 #define QCA_WCN3991_SOC_ID		(0x40014320)
+#define QCA_QCA6696_SOC_ID		(0x400A0200)
+#define QCA_QCA6595_SOC_ID		(0x400B0200)
+#define QCA_QCA6698_SOC_ID		(0x400C0210)
+#define QCA_QCA6797_SOC_ID		(0x40170200)
 
 /* QCA chipset version can be decided by patch and SoC
  * version, combination with upper 2 bytes from SoC
@@ -154,6 +158,7 @@ enum qca_btsoc_type {
 	QCA_WCN6750,
 	QCA_WCN6855,
 	QCA_WCN7850,
+	QCA_AUTO_CONVERGED,
 };
 
 #if IS_ENABLED(CONFIG_BT_QCA)
diff --git a/drivers/bluetooth/hci_qca.c b/drivers/bluetooth/hci_qca.c
index 3870834976a9..ec7d1176576f 100644
--- a/drivers/bluetooth/hci_qca.c
+++ b/drivers/bluetooth/hci_qca.c
@@ -1384,6 +1384,7 @@ static int qca_set_baudrate(struct hci_dev *hdev, uint8_t baudrate)
 	case QCA_WCN6750:
 	case QCA_WCN6855:
 	case QCA_WCN7850:
+	case QCA_AUTO_CONVERGED:
 		msleep(30);
 		break;
 
@@ -1470,6 +1471,7 @@ static int qca_check_speeds(struct hci_uart *hu)
 	case QCA_WCN6750:
 	case QCA_WCN6855:
 	case QCA_WCN7850:
+	case QCA_AUTO_CONVERGED:
 		if (!qca_get_speed(hu, QCA_INIT_SPEED) &&
 		    !qca_get_speed(hu, QCA_OPER_SPEED))
 			return -EINVAL;
@@ -1512,6 +1514,7 @@ static int qca_set_speed(struct hci_uart *hu, enum qca_speed_type speed_type)
 		case QCA_WCN6750:
 		case QCA_WCN6855:
 		case QCA_WCN7850:
+		case QCA_AUTO_CONVERGED:
 			hci_uart_set_flow_control(hu, true);
 			break;
 
@@ -1549,6 +1552,7 @@ static int qca_set_speed(struct hci_uart *hu, enum qca_speed_type speed_type)
 		case QCA_WCN6750:
 		case QCA_WCN6855:
 		case QCA_WCN7850:
+		case QCA_AUTO_CONVERGED:
 			hci_uart_set_flow_control(hu, false);
 			break;
 
@@ -1926,6 +1930,9 @@ static int qca_setup(struct hci_uart *hu)
 	case QCA_WCN7850:
 		soc_name = "wcn7850";
 		break;
+	case QCA_AUTO_CONVERGED:
+		soc_name = "qca_auto";
+		break;
 
 	default:
 		soc_name = "ROME/QCA6390";
@@ -1949,6 +1956,7 @@ static int qca_setup(struct hci_uart *hu)
 	case QCA_WCN6750:
 	case QCA_WCN6855:
 	case QCA_WCN7850:
+	case QCA_AUTO_CONVERGED:
 		qcadev = serdev_device_get_drvdata(hu->serdev);
 		if (qcadev->bdaddr_property_broken)
 			set_bit(HCI_QUIRK_BDADDR_PROPERTY_BROKEN, &hdev->quirks);
@@ -1981,6 +1989,7 @@ static int qca_setup(struct hci_uart *hu)
 	case QCA_WCN6750:
 	case QCA_WCN6855:
 	case QCA_WCN7850:
+	case QCA_AUTO_CONVERGED:
 		break;
 
 	default:
@@ -1994,7 +2003,9 @@ static int qca_setup(struct hci_uart *hu)
 	ret = qca_uart_setup(hdev, qca_baudrate, soc_type, ver,
 			firmware_name, rampatch_name);
 	if (!ret) {
-		clear_bit(QCA_IBS_DISABLED, &qca->flags);
+		if (soc_type != QCA_AUTO_CONVERGED)
+			clear_bit(QCA_IBS_DISABLED, &qca->flags);
+
 		qca_debugfs_init(hdev);
 		hu->hdev->hw_error = qca_hw_error;
 		hu->hdev->cmd_timeout = qca_cmd_timeout;
@@ -2159,6 +2170,12 @@ static const struct qca_device_data qca_soc_data_wcn7850 __maybe_unused = {
 	.capabilities = QCA_CAP_WIDEBAND_SPEECH | QCA_CAP_VALID_LE_STATES,
 };
 
+static const struct qca_device_data qca_soc_data_auto __maybe_unused = {
+	.soc_type = QCA_AUTO_CONVERGED,
+	.num_vregs = 0,
+	.capabilities = QCA_CAP_WIDEBAND_SPEECH | QCA_CAP_VALID_LE_STATES,
+};
+
 static void qca_power_shutdown(struct hci_uart *hu)
 {
 	struct qca_serdev *qcadev;
@@ -2398,6 +2415,15 @@ static int qca_serdev_probe(struct serdev_device *serdev)
 		}
 		break;
 
+	case QCA_AUTO_CONVERGED:
+		power_ctrl_enabled = false;
+		err = hci_uart_register_device(&qcadev->serdev_hu, &qca_proto);
+		if (err) {
+			BT_ERR("qca auto serdev registration failed");
+			return err;
+		}
+		break;
+
 	default:
 		qcadev->bt_en = devm_gpiod_get_optional(&serdev->dev, "enable",
 					       GPIOD_OUT_LOW);
@@ -2678,6 +2704,7 @@ static const struct of_device_id qca_bluetooth_of_match[] = {
 	{ .compatible = "qcom,wcn6750-bt", .data = &qca_soc_data_wcn6750},
 	{ .compatible = "qcom,wcn6855-bt", .data = &qca_soc_data_wcn6855},
 	{ .compatible = "qcom,wcn7850-bt", .data = &qca_soc_data_wcn7850},
+	{ .compatible = "qcom,auto-converged-bt", .data = &qca_soc_data_auto},
 	{ /* sentinel */ }
 };
 MODULE_DEVICE_TABLE(of, qca_bluetooth_of_match);
-- 
2.34.1

